lib/jekyll/reader.rb

```diff
@@ -16,9 +16,19 @@ def read
       read_directories
       read_included_excludes
       sort_files!
-      @site.data = DataReader.new(site).read(site.config["data_dir"])
       CollectionReader.new(site).read
       ThemeAssetsReader.new(site).read
+      read_data
+    end
+
+    # Read and merge the contents of _data folders. Data from Site takes
+    # precedence over data from theme.
+    #
+    # Returns nothing.
+    def read_data
+      @site.data = DataReader.new(site).read(site.config["data_dir"])
+      theme_data = DataReader.new(site).read(site.theme&.data_path)
+      Jekyll::Utils.deep_merge_hashes(theme_data, @site.data)
```

Comment:
@mgerzabek I doubt this works like you expect it to..
`@site.data` isn't being mutated.

You should either reassign `@site.data` or use `Utils.deep_merge_hashes!`

I don't see tests.
You can modify the local `test-theme` for this and add tests (similar to ones existing for layouts, includes and assets in a theme).

Reply:
Okay, @ashmaroli sorry for the noise… assigned the merged hashes now in read method.

> I don't see tests.

I’m (not so) new to Ruby (anymore), but have no idea on how to write tests. If you could point me to a good primer to testing in Ruby especially in context of Jekyll, I’d be very thankful. I’ll look into testing over the weekend.

Reply:
@mgerzabek This will need two types of tests.
But first update our [test-theme](https://github.com/jekyll/jekyll/tree/master/test/fixtures/test-theme).
Then try reading through this [unit test file](https://github.com/jekyll/jekyll/blob/master/test/test_theme_assets_reader.rb) and extrapolate for data files in theme.
Finally, add scenarios to this [Cucumber feature](https://github.com/jekyll/jekyll/blob/master/features/theme.feature) (more important than adding a unit test for this pull request).